src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -0,0 +1,298 @@
+  auto result = FreeInTracee(pid, address_code, size);
+  FAIL_IF(result.has_error(), "Unable to free previously allocated memory in tracee: \"%s\"",
+          result.error().message());
+}
+
+void RestoreRegistersOrDie(RegisterState& register_state) {
+  auto result = register_state.RestoreRegisters();
+  FAIL_IF(result.has_error(), "Unable to restore register state in tracee: \"%s\"",
+          result.error().message());
+}
+
+uint64_t GetReturnValueOrDie(pid_t pid) {
+  RegisterState return_value_registers;
+  auto result_return_value = return_value_registers.BackupRegisters(pid);
+  FAIL_IF(result_return_value.has_error(), "Unable to read registers after function called :\"%s\"",
+          result_return_value.error().message());
+  return return_value_registers.GetGeneralPurposeRegisters()->x86_64.rax;
+}
+
+// Execute the code at `address_code`. The code segment has to end with an `int3`.
+void ExecuteOrDie(pid_t pid, RegisterState& original_registers, uint64_t address_code) {
+  RegisterState registers_set_rip = original_registers;
+  registers_set_rip.GetGeneralPurposeRegisters()->x86_64.rip = address_code;
+  auto result_restore_registers = registers_set_rip.RestoreRegisters();
+  if (result_restore_registers.has_error()) {
+    FATAL("Unable to continue tracee with PTRACE_CONT.");
+  }
+  if (ptrace(PTRACE_CONT, pid, 0, 0) != 0) {
+    FATAL("Unable to continue tracee with PTRACE_CONT.");
+  }
+  int status = 0;
+  pid_t waited = waitpid(pid, &status, 0);
+  if (waited != pid || !WIFSTOPPED(status) || WSTOPSIG(status) != SIGTRAP) {
+    FATAL("Failed to wait for sigtrap after PTRACE_CONT.");
+  }
+}
+
+}  // namespace
+
+[[nodiscard]] ErrorMessageOr<void*> DlopenInTracee(pid_t pid, std::filesystem::path path,
+                                                   uint32_t flag) {
+  // Figure out address of dlopen in libc.
+  OUTCOME_TRY(address_dlopen, FindFunctionAddress(pid, "dlopen", "libc-"));
+
+  // Backup registers.
+  RegisterState original_registers;
+  OUTCOME_TRY(original_registers.BackupRegisters(pid));
+
+  // Allocate small memory area in the tracee. This is used for the code and the path name.
+  const uint64_t path_length = path.string().length() + 1;  // Include terminating zero.
+  const uint64_t memory_size = kScratchPadSize + path_length;
+  OUTCOME_TRY(address_code, AllocateInTracee(pid, memory_size));
+
+  // Write the name of the .so into memory at address_code with offset of kScratchPadSize.
+  std::vector<uint8_t> path_as_vector(path_length);
+  memcpy(path_as_vector.data(), path.c_str(), path_length);
+  const uint64_t address_so_path = address_code + kScratchPadSize;
+  auto result_write_path = WriteTraceesMemory(pid, address_so_path, path_as_vector);
+  if (result_write_path.has_error()) {
+    FreeMemoryOrDie(pid, address_code, memory_size);
+    return result_write_path.error();
+  }
+  // We want to do the following in the tracee:
+  // return_value = dlopen(path, RTLD_LAZY);
```

Comment:
nit: use RTLD_NOW

Reply:
Changed the comment to what actually happens in the code: we forward the flag parameter here. 

Changed to using RTLD_NOW in the test as discussed offline.